toolchain/check/type_completion.cpp

```diff
@@ -872,7 +921,12 @@ auto RequireIdentifiedFacetType(Context& context,
       for (auto require_impls_id : context.require_impls_blocks().Get(
                constraint.require_impls_block_id)) {
         const auto& require = context.require_impls().Get(require_impls_id);
-        impls_facet_types.push_back(require.facet_type_id);
+        auto require_specific_id = MakeCopyOfSpecificAndAppendSelf(
+            context, loc_id, impls.specific_id, require.generic_id,
+            GetRequireImplsSpecificSelf(context, require));
+        auto facet_type_id = GetFacetTypeInSpecific(
+            context, require.facet_type_inst_id, require_specific_id);
```

Comment:
The way these are composed, it seems like the APIs are intended to be generic, but it seems like (for example) `MakeCopyOfSpecificAndAppendSelf` and `GetRequireImplsSpecificSelf` are tightly associated in their assumption that the `Self` will be the last parameter. Had you considered an API that took in `require` and returned `facet_type_id`?

e.g. `auto facet_type_id = GetFacetTypeInRequireImpls(context, loc_id, impls.specific_id, require);`

Reply:
Yeah I was just trying to break things into smaller pieces. I have been thinking about how I will need to do this same operation in more places. In particular, in an `impl` declaration, we will need to find all the required interfaces (in the impl's specific facet type) and ensure they have been implemented already. So I think this will need to become a single operation that is exposed elsewhere, probably from `sem_ir/require_impls.h`. But for now, leaving it here.

I can combine some of these steps.